src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerGenotypingEngine.java

```diff
@@ -111,24 +112,98 @@ public CalledHaplotypes(final List<VariantContext> calls, final Set<Haplotype> c
+    }
+
+    /**
+     * Returns the list of variant contexts active at this location, using either the list of active haplotypes or
+     * the list of active alleles to genotypes if we are in GGA mode as candidates. This method will include events
+     * that span the current location if includeSpanningEvents is set to true; otherwise it will only include events that
+     * have loc as their start position.
+     * @param haplotypes Haplotypes for the current active region.
+     * @param loc The start position we are genotyping
+     * @param activeAllelesToGenotype The list of given alleles for the current active region, empty unless we are in GGA mode
+     * @param includeSpanningEvents If true, will also return events that span loc
+     */
+    protected static List<VariantContext> getVCsAtThisLocation(final List<Haplotype> haplotypes,
+                                                               final int loc,
+                                                               final List<VariantContext> activeAllelesToGenotype,
+                                                               final boolean includeSpanningEvents) {
         // the overlapping events to merge into a common reference view
-        final List<VariantContext> eventsAtThisLoc = new ArrayList<>();
+        final List<VariantContext> results = new ArrayList<>();
+        final Set<LocationAndAlleles> uniqueLocationsAndAlleles = new HashSet<>();
 
         if( activeAllelesToGenotype.isEmpty() ) {
-            final List<VariantContext> events = haplotypes.stream().map(h -> h.getEventMap().get(loc))
-                    .filter(Objects::nonNull).collect(Collectors.toList());
-            for (final VariantContext vc : events) {
-                if (eventsAtThisLoc.stream().noneMatch(vc::hasSameAllelesAs)) {
-                    eventsAtThisLoc.add(vc);
-                }
-            }
+            haplotypes.stream()
+                    .flatMap(h -> Utils.stream(h.getEventMap().getSpanningEvents(loc)))
+                    .filter(Objects::nonNull)
+                    .filter(v -> (includeSpanningEvents || v.getStart() == loc))
+                    .forEach(v -> {
+                        final LocationAndAlleles locationAndAlleles = new LocationAndAlleles(v.getStart(), v.getAlleles());
+                        if (! uniqueLocationsAndAlleles.contains(locationAndAlleles)) {
+                            uniqueLocationsAndAlleles.add(locationAndAlleles);
+                            results.add(v);
+                        }
+                    });
         } else { // we are in GGA mode!
             int compCount = 0;
             for( final VariantContext compVC : activeAllelesToGenotype ) {
-                if( compVC.getStart() == loc ) {
+                if( compVC.getStart() <= loc && compVC.getEnd() >= loc) {
+                    if (! (includeSpanningEvents || compVC.getStart() == loc)) {
```

Comment:
Passing on this suggestion for now since I feel like it wouldn't be much clearer to my eye. Let me know if you really want me to re-write it that way.

Reply:
Fine with me.